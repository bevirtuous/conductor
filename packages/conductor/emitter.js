module.exports = require('./src/emitter');
