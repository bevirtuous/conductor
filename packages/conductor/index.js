module.exports = require('./src/Conductor');
