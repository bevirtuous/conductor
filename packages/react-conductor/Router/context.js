module.exports = require('../src/context');