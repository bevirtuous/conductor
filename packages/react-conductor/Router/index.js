const { RouteContext } = require('../src/context');
module.exports = require('../src/Router');
module.exports.RouteContext = RouteContext;