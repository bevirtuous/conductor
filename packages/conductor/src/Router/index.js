import uuid from 'uuid/v4';
import Route from '../Route';
import * as errors from './errors';
import emitter from '../emitter';
import history from '../history';
import matcher from '../matcher';
import stack from '../Stack';
import * as constants from '../constants';

/**
 * The Router class.
 */
class Router {
  /**
   * TODO:
   */
  constructor(createHistory = history) {
    // Flag to indicate a native history event. Should always be reset to true.
    this.nativeEvent = true;

    this.history = createHistory();

    // The patterns are collected to match against pathnames.
    this.patterns = {};

    // Flag to indicate whether or not there is an ongoing history change.
    this.routing = false;

    // The `routeIndex` is used to track which stack entry is the current route.
    this.routeIndex = 0;

    this.action = constants.ACTION_PUSH;

    // Unsubscribe to any other history module changes.
    if (typeof this.historyListener === 'function') {
      this.historyListener();
      stack.clear();
    }

    // 
    this.addInitialRoute();

    this.historyListener = this.history.listen(this.handleNativeEvent);
  }

  /**
   * 
   */
  handleNativeEvent = (location, action) => {
    if (!this.nativeEvent) {
      return;
    }

    const next = stack.getByIndex(this.routeIndex + 1);

    if (next && location.state && location.state.route && next.id === location.state.route.id) {
      this.handlePush({
        pathname: location.pathname,
        state: location.state,
      }, false);

      return;
    }

    if (action === constants.ACTION_POP) {
      this.handlePop();
    }
  }

  /**
   * Populate the stack with an initial entry to match the history module.
   * Note: we cannot match it against a pattern at this point.
   */
  addInitialRoute = () => {
    const { hash, pathname, search } = this.history.location;
    const fullPathname = `${pathname}${search}${hash}`;
    const id = uuid();
    stack.add(id, new Route({ id, pathname: fullPathname }));
  }

  /**
   * @param {Object} params The router action params.
   * @returns {Promise}
   */
  handlePop = (params = {}) => {
    return new Promise((resolve, reject) => {
      const {
        emitBefore = true,
        emitAfter = true,
        steps = 1,
        state = null,
      } = params;
      let unlisten = null;
      const { size } = stack.getAll();

      if (size < 2) {
        reject(new Error(errors.ESTACKLENGTH));
        this.nativeEvent = true;
        return;
      }

      if (steps <= 0) {
        reject(new Error(errors.EINVALIDSTEPS));
        this.nativeEvent = true;
        return;
      }

      // Get id of target route.
      const targetIndex = Math.max(this.routeIndex - steps, 0);
      const prev = stack.getByIndex(this.routeIndex);
      const next = stack.getByIndex(targetIndex);
      const end = { prev, next };

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_POP, end);
      }

      if (state) {
        next.state = Object.assign(next.state, state);
      }

      /**
       * 
       */
      const callback = () => {
        unlisten();
        this.routeIndex = targetIndex;
        this.action = constants.ACTION_POP;

        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_POP, end);
        }

        resolve(end);
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history back action.
      if (!this.nativeEvent) {
        this.history.go(steps * -1);
      } else {
        callback();
      }
    });
  }

  /**
   * @param {Object} params The params to use when navigating.
   * @returns {Promise}
   */
  handlePush(params, override = true) {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      // Check for ongoing router action.
      if (this.routing) {
        // TODO: create real error
        reject(new Error('Error'));
        this.nativeEvent = true;
        return;
      }

      const {
        emitBefore = true,
        emitAfter = true,
        pathname,
        state,
      } = params;
      const pattern = this.findPattern(pathname.split('?')[0]);
      let unlisten = null;

      if (!pattern) {
        reject(new Error(errors.EINVALIDPATHNAME));
        this.nativeEvent = true;
        return;
      }

      // Block further router actions until this Promise has been returned.
      this.routing = true;

      // Remove all unwanted items from the stack.
      if (override) {
        while (this.routeIndex < stack.getAll().size - 1) {
          const [id] = stack.last();
          stack.remove(id);
        }
      }

      const id = uuid();
      const { transform } = this.patterns[pattern];
      const prev = stack.getByIndex(this.routeIndex);
      const next = new Route({
        id,
        pathname,
        pattern,
        state,
        transform,
      });

      // Add item to the stack
      stack.add(id, next);

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_PUSH, { prev, next });
      }

      /**
       * TODO: move to class function with params
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        // Increment the route index.
        this.routeIndex += 1;

        this.action = constants.ACTION_PUSH;

        // Emit completion event.
        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_PUSH, { prev, next });
        }

        // Resolve the Promise.
        resolve({ prev, next });
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history push action.
      if (!this.nativeEvent) {
        this.history.push({
          pathname,
          state: {
            ...state,
            route: { id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * Match the given pathname to a registered pattern.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  findPattern = (pathname) => {
    const pattern = Object.keys(this.patterns).find(key => this.patterns[key].match(pathname));
    return pattern || null;
  }

  /**
   * Registers a route pattern to match new pathnames against.
   * @param {string} pattern The pattern to register.
   * @param {Function} transform The route transformer.
   */
  // TODO: convert to Promise
  register = (pattern, transform = null) => {
    if (!pattern) {
      throw new Error(errors.EMISSINGPATTERN);
    }

    if (typeof pattern !== 'string') {
      throw new Error(errors.EINVALIDPATTERN);
    }

    // 
    const match = matcher(pattern);

    // 
    this.patterns[pattern] = {
      match,
      transform,
    };

    // Find the pathname of the first route.
    const [, route] = stack.first();

    // If it has been set then we don't need to match it.
    if (route.pattern !== null) {
      return;
    }

    // 
    if (match(route.pathname)) {
      route.setPattern(pattern);

      const end = { prev: null, next: route };

      emitter.emit(constants.EVENT_WILL_PUSH, end);
      emitter.emit(constants.EVENT_DID_PUSH, end);
    }
  }

  handleReplace = (params) => {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      // Check for ongoing router action.
      if (this.routing) {
        // TODO: create real error
        reject(new Error('Error'));
        this.nativeEvent = true;
        return;
      }

      const {
        emitBefore = true,
        emitAfter = true,
        pathname,
        state,
      } = params;
      const pattern = this.findPattern(pathname.split('?')[0]);
      let unlisten = null;

      if (!pattern) {
        reject(new Error(errors.EINVALIDPATHNAME));
        this.nativeEvent = true;
        return;
      }

      // Block further router actions until this Promise has been returned.
      this.routing = true;

      const { id } = stack.getByIndex(this.routeIndex);
      const { transform } = this.patterns[pattern];
      const prev = stack.get(id);
      const next = new Route({
        id,
        pathname,
        pattern,
        state,
        transform,
      });
      const end = { prev, next };

      // Add item to the stack
      stack.add(id, next);

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_REPLACE, end);
      }

      /**
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        this.action = constants.ACTION_REPLACE;

        // Emit completion event.
        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_REPLACE, end);
        }

        resolve(end);
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history replace action.
      if (!this.nativeEvent) {
        this.history.replace({
          pathname,
          state: {
            ...state,
            route: { id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  push = (params) => {
    this.nativeEvent = false;
    return this.handlePush(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  pop = (params) => {
    this.nativeEvent = false;
    return this.handlePop(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  replace = (params) => {
    this.nativeEvent = false;
    return this.handleReplace(params);
  }

  /**
   * 
   */
  reset = () => {
    return new Promise((resolve, reject) => {
      const { size } = stack.getAll();
      const [, route] = stack.first();

      const prev = stack.getByIndex(this.routeIndex);
      const end = {
        prev,
        next: route,
      };

      if (size === 1) {
        // TODO: Add error
        reject();
        return;
      }

      emitter.emit(constants.EVENT_WILL_RESET, end);

      this.pop({
        emitBefore: false,
        emitAfter: false,
        steps: size,
      }).then(() => {
        emitter.emit(constants.EVENT_DID_RESET, end);
        resolve(end);
      });
    });
  }

  /**
   * @param {string} pathname The pathname to reset to.
   * @returns {Promise}
   */
  resetTo = (pathname) => {
    return new Promise((resolve, reject) => {
      if (!pathname) {
        reject(new Error(errors.EMISSINGPATHNAME));
        return;
      }

      if (!this.findPattern(pathname)) {
        reject(new Error(errors.EINVALIDPATHNAME));
        return;
      }

      const id = uuid();
      const prev = stack.getByIndex(this.routeIndex);
      const next = new Route({ id, pathname });
      const end = { prev, next };

      emitter.emit(constants.EVENT_WILL_RESET, end);
      stack.reset([id, next]);
      emitter.emit(constants.EVENT_DID_RESET, end);

      resolve(end);
    });
  }

  /**
   * 
   */
  update = (id, state = {}, emit = true) => {
    return new Promise((resolve, reject) => {
      //
      if (!id || Object.keys(state).length === 0) {
        reject(new Error(errors.EPARAMSINVALID));
        return;
      }

      const route = stack.get(id);

      if (!route) {
        reject(new Error(errors.EINVALIDID));
        return;
      }

      route.state = {
        ...route.state,
        ...state,
      };
      route.updated = Date.now();

      stack.update(id, route);

      if (emit) {
        emitter.emit(constants.EVENT_UPDATE, route);
      }

      resolve(route);
    });
  }

  // TODO: Add deprecation warning
  getCurrentRoute = () => (
    stack.getByIndex(this.routeIndex)
  )
}

export default new Router();
