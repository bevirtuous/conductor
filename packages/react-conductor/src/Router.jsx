import React, { Component } from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import * as events from '@virtuous/conductor-events';
import getRouteStack from '@virtuous/conductor-helpers/src/getRouteStack';

/**
 * The Router component.
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    initialPath: PropTypes.string.isRequired,
    initialPattern: PropTypes.string.isRequired,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      routeStack: [{
        path: props.initialPath,
        pattern: props.initialPattern,
      }],
    };

    // Setup some listeners for router events.
    events.onPush(this.handleRouteChange);
    events.onPop(this.handleRouteChange);
    events.onReplace(this.handleRouteChange);

    // Push the first route.
    setTimeout(() => {
      conductor.push(props.initialPath);
    }, 0);
  }

  /**
   * Limit re-rendering unless the state actually changes.
   *
   * @param {Object} nextProps The next props.
   * @param {Object} nextState The next state.
   * @returns {boolean}
   */
  shouldComponentUpdate() {
    // TODO: Add some sort of check.
    return true;
  }

  /**
   * Find the index of the given route pattern against the currently open routes.
   * @param {string} pattern The pattern to check for.
   * @returns {number} The index of the pattern.
   */
  getRouteIndex(pattern) {
    for (let i = 0, j = this.state.routeStack.length; i < j; i += 1) {
      if (this.state.routeStack[i].pattern === pattern) {
        return i;
      }
    }
    return -1;
  }

  /**
   * Update the state based on the router changes.
   * @param {string} action The history action performed.
   * @param {string} currentPath The newly opened path.
   * @param {string} prevPath The previous path.
   * @param {string} routeStack The open routes.
   */
  handleRouteChange = () => {
    const routeStack = getRouteStack();
    this.setState({ routeStack });
  }

  /**
   * The render method.
   * @returns {JSX}
   */
  render() {
    /**
     * We need to apply some props to the children based on
     * the current router state.
     */
    // TODO: Move this code to another function.
    return React.Children.map(this.props.children, (route, index) => {

      // TODO: This is not the correct way to get the index.
      // Get the index of this route.
      const routeIndex = this.getRouteIndex(route.props.pattern);

      // Check if this route is open.
      const isOpen = routeIndex > -1;

      // Check if this route is the current visible route.
      const isVisible = routeIndex === this.state.routeStack.length - 1;

      // If this route is open then pass the pathname as a prop.
      const path = isOpen ? this.state.routeStack[routeIndex].pathname : null;
      const state = isOpen ? this.state.routeStack[routeIndex].state : null;

      return React.cloneElement(route, {
        index,
        isOpen,
        isVisible,
        path,
        state,
      });
    });
  }
}

export default Router;
