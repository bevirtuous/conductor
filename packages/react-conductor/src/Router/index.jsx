import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import * as constants from '@virtuous/conductor/constants';
import * as events from '@virtuous/conductor-events';
import getCurrentAction from '@virtuous/conductor-helpers/getCurrentAction';
import getRouteStack from '@virtuous/conductor-helpers/getRouteStack';
import Route from '../Route';

const Context = React.createContext();
export const RouteContext = Context;

/**
 * The Router component.
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      stack: [],
    };

    // The route definitions.
    this.routes = {};

    // The stack of route definitions to use when rendering.
    this.routeStack = [];

    // Register the routes with Conductor.
    React.Children.forEach(props.children, child => this.registerChild(child.props));

    // Setup the callbacks for routing events.
    events.onDidPush(() => this.handleRouteChange(constants.ACTION_PUSH));
    events.onDidPop(() => this.handleRouteChange(constants.ACTION_POP));
    events.onDidReplace(() => this.handleRouteChange(constants.ACTION_REPLACE));
    events.onDidReset(() => this.handleRouteChange(constants.ACTION_RESET));
  }

  /**
   * @param {Object} nextProps The next props.
   * @param {Object} nextState The next state.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps, nextState) {
    // Render if the stacks are different sizes.
    if (this.state.stack.length !== nextState.stack.length) {
      return true;
    }

    // Render if the history action was a replace.
    if (getCurrentAction() === constants.ACTION_REPLACE) {
      return true;
    }

    return false;
  }

  /**
   * Determines the last occurence of a path in the stack of open routes.
   * @param {Array} stack The open routes stack.
   * @param {string} pattern The pattern to look for.
   * @returns {number|null}
   */
  getlastOccurence = (stack, pattern) => {
    for (let i = stack.length - 1; i >= 0; --i) {
      if (stack[i].pattern === pattern) {
        return i;
      }
    }

    return null;
  }

  /**
   * Update the state based on the router changes.
   * TODO:
   */
  handleRouteChange = (action) => {
    const stack = getRouteStack();

    switch (action) {
      case constants.ACTION_PUSH: {
        //
        const { pattern } = stack[stack.length - 1];

        //
        if (this.routes[pattern].preload) {
          this.routeStack.push(this.routes[pattern].pattern);
        }

        break;
      }

      case constants.ACTION_POP: {
        //
        const { pattern } = this.state.stack[this.state.stack.length - 1];

        //
        if (this.routes[pattern].preload) {
          this.routeStack.pop();
        }

        break;
      }

      case constants.ACTION_REPLACE: {
        const { pattern } = this.state.stack[this.state.stack.length - 1];
        const newPattern = stack[stack.length - 1].pattern;

        //
        if (this.routes[pattern].preload) {
          this.routeStack.pop();
        }

        //
        if (this.routes[newPattern].preload) {
          this.routeStack.push(this.routes[newPattern].pattern);
        }

        break;
      }

      case constants.ACTION_RESET: {
        const { pattern, preload } = this.state.stack[0];

        // Set the routeStack back to it's initial state.
        this.routeStack.length = this.props.children.length;

        /**
         * Check if the initial entry is the stack is a preload route.
         * If so then immediately push another to the route stack.
         */
        if (preload) {
          this.routeStack.push(pattern);
        }

        break;
      }

      default:
        break;
    }

    this.setState({ stack });
  };

  /**
   * @param {Object} props - The child's props.
   */
  registerChild(props) {
    conductor.register(props.pattern);

    this.routes[props.pattern] = {
      component: props.component,
      pattern: props.pattern,
      preload: props.preload,
    };

    this.routeStack.push(props.pattern);
  }

  /**
   * TODO:
   * @returns {JSX}
   */
  renderOpenRoutes() {
    const ignoredSingletons = [];
    const usedRoutes = [];

    return (
      <Fragment>
        {this.routeStack.map((item, index) => {
          let pathname = null;
          let isVisible = false;
          let id = null;
          let params = {};
          let query = {};
          let state = {};

          const { component, pattern, preload } = this.routes[item];

          if (preload) {
            const match = this.state.stack.find((stackItem, stackItemIndex) => {
              const isMatch =
                stackItem.pattern === pattern && !usedRoutes.includes(stackItemIndex);

              if (isMatch) {
                usedRoutes.push(stackItemIndex);
              }

              if (isMatch && stackItemIndex === this.state.stack.length - 1) {
                isVisible = true;
              }

              return isMatch;
            });

            if (match) {
              id = match.id;
              pathname = match.pathname;
              params = match.params;
              query = match.query;
              state = match.state;
            }
          } else {
            // Skip this route if this pattern has already been handled.
            if (ignoredSingletons.includes(pattern)) {
              return null;
            }

            // We are handling this pattern only once, ignore it.
            ignoredSingletons.push(pattern);

            const lastOccurence = this.getlastOccurence(
              this.state.stack,
              pattern
            );

            if (this.state.stack[lastOccurence]) {
              id = this.state.stack[lastOccurence].id;
              pathname = this.state.stack[lastOccurence].pathname;
              params = this.state.stack[lastOccurence].params;
              query = this.state.stack[lastOccurence].query;
              state = this.state.stack[lastOccurence].state;
            }

            isVisible = lastOccurence === this.state.stack.length - 1;
          }

          const key = `route-${index}`;

          return (
            <Route
              id={id}
              component={component}
              isVisible={isVisible}
              key={key}
              params={params}
              path={pathname}
              pattern={pattern}
              query={query}
              state={state}
            />
          );
        })}
      </Fragment>
    );
  }

  /**
   * The render method.
   * @returns {JSX}
   */
  render() {
    return this.renderOpenRoutes();
  }
}

export default Router;
