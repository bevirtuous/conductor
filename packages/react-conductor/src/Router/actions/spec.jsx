import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { router } from '@virtuous/conductor';
import Router from '../index';

describe('<Router.{Action} />', () => {
  it('should render as expected', () => {
    const className = 'sausage';
    const state = {
      something: true,
    };

    const pushSpy = jest.spyOn(router, 'push');
    const popSpy = jest.spyOn(router, 'pop');
    const replaceSpy = jest.spyOn(router, 'replace');
    const resetSpy = jest.spyOn(router, 'reset');
    const resetToSpy = jest.spyOn(router, 'resetTo');

    const app = render((
      <Router>
        <Router.Push className={className} to="/" state={state}>
          Push!
        </Router.Push>
        <Router.Pop className={className} steps={2} state={state}>
          Pop!
        </Router.Pop>
        <Router.Replace className={className} to="/" state={state}>
          Replace!
        </Router.Replace>
        <Router.Reset className={className}>
          Reset!
        </Router.Reset>
        <Router.ResetTo className={className} to="/" state={state}>
          Reset to!
        </Router.ResetTo>
      </Router>
    ));

    fireEvent.click(app.queryByText('Push!'));
    fireEvent.click(app.queryByText('Pop!'));
    fireEvent.click(app.queryByText('Replace!'));
    fireEvent.click(app.queryByText('Reset!'));
    fireEvent.click(app.queryByText('Reset to!'));

    expect(app).toMatchSnapshot();
    expect(pushSpy).toHaveBeenCalled();
    expect(popSpy).toHaveBeenCalled();
    expect(replaceSpy).toHaveBeenCalled();
    expect(resetSpy).toHaveBeenCalled();
    expect(resetToSpy).toHaveBeenCalled();
  });
});
