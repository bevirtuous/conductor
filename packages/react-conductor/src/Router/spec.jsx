import React from 'react';
import { shallow, mount } from 'enzyme';
import { router } from '@virtuous/conductor';
import Router from './index';

const spy = jest.spyOn(router, 'constructor');

describe('<Router />', () => {
  it('should render with children and context', () => {
    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    const child = app.find('p');
    expect(child.length).toBe(1);
    expect(app).toMatchSnapshot();

    // TODO: Test Router context value
  });

  it('should re-render with children and context', () => {
    // TODO: Test Router context value
  });

  it.only('should pass the history function to core', () => {
    const mockHistory = () => ({
      listen: () => {},
      location: {},
    });

    shallow((
      <Router history={mockHistory}>
        <p />
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(mockHistory);
  });
});
