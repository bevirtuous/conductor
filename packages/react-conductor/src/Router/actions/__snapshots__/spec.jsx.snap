// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Router.{Action} /> should render as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="sausage"
        href="/"
      >
        Push!
      </a>
      <a
        class="sausage"
      >
        Pop!
      </a>
      <a
        class="sausage"
        href="/"
      >
        Replace!
      </a>
      <a
        class="sausage"
      >
        Reset!
      </a>
      <a
        class="sausage"
        href="/"
      >
        Reset to!
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="sausage"
      href="/"
    >
      Push!
    </a>
    <a
      class="sausage"
    >
      Pop!
    </a>
    <a
      class="sausage"
      href="/"
    >
      Replace!
    </a>
    <a
      class="sausage"
    >
      Reset!
    </a>
    <a
      class="sausage"
      href="/"
    >
      Reset to!
    </a>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
